### PR TITLE
Move e2e tests for metrics export to integration status to reduce crosstalk flakiness

### LINKED ISSUE #1672: TestMetricsExport is super flaky
metrics/resource_view_test.go - TestMetricsExport flakes most of the time.

/assign @jjzeng-seattle @evankanderson 

here's a good example: https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative_pkg/1666/pull-knative-pkg-unit-tests/1300475037081407489

### PR DESCRIPTION
**Pull request by evankanderson**
Maybe fixes #1672 

Creating a PR to see what I need to change in Prow

/hold

### CODE REVIEW

File: metrics/record_e2e_test.go [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+	"time"
+
+	sd "contrib.go.opencensus.io/exporter/stackdriver"
+	ocmetrics "github.com/census-instrumentation/opencensus-proto/gen-go/agent/metrics/v1"
+	ocresource "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	"go.opencensus.io/resource"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"k8s.io/apimachinery/pkg/util/clock"
```
Comment (mattmoor) [inline metrics/record_e2e_test.go:47]: Format Go code:
```suggestion

```

File: metrics/resource_view_test.go [hunk 2]
```diff
@@ -17,47 +17,18 @@
...
-	"google.golang.org/grpc"
-	proto "google.golang.org/protobuf/proto"
-
-	logtesting "knative.dev/pkg/logging/testing"
-	"knative.dev/pkg/metrics/metricskey"
-	"knative.dev/pkg/metrics/metricstest"
 )
 
 var (
 	r               = resource.Resource{Labels: map[string]string{"foo": "bar"}}
```
Comment (mattmoor) [inline metrics/resource_view_test.go:31]: Format Go code:
```suggestion
	r = resource.Resource{Labels: map[string]string{"foo": "bar"}}
```

**Review by mattmoor — COMMENTED** [review]
Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -path './third_party' -prune -o -name '*.pb.go' -prune -o -type f -name '*.go' -print)`
  `goimports -w $(find -name '*.go' | grep -v vendor | grep -v third_party | grep -v .pb.go | grep -v wire_gen.go)`

File: metrics/record_e2e_test.go [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+			resp, err := http.Get(fmt.Sprintf("http://localhost:%d/metrics", prometheusPort))
+			if err != nil {
+				t.Fatalf("failed to fetch prometheus metrics: %+v", err)
+			}
+			defer resp.Body.Close()
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				t.Fatalf("failed to read prometheus response: %+v", err)
+			}
+			want := `# HELP testComponent_global_export_counts Count of exports via standard OpenCensus view.
```
Comment (vagababov) [inline metrics/record_e2e_test.go:213]:
```suggestion
			const want = `# HELP testComponent_global_export_counts Count of exports via standard OpenCensus view.
```

File: metrics/record_e2e_test.go [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+	"knative.dev/pkg/metrics/metricskey"
+	"knative.dev/pkg/metrics/metricstest"
+)
+
+var (
+	NamespaceTagKey = tag.MustNewKey(metricskey.LabelNamespaceName)
+	ServiceTagKey   = tag.MustNewKey(metricskey.LabelServiceName)
+	ConfigTagKey    = tag.MustNewKey(metricskey.LabelConfigurationName)
+	RevisionTagKey  = tag.MustNewKey(metricskey.LabelRevisionName)
+)
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply metrics/record_e2e_test.go:66]: I think making them package-private is probably more appropriate than trying to share them. Done.

File: metrics/record_e2e_test.go [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+		return err
+	}
+	os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", tmp.Name())
+	return nil
+}
+
+func sortMetrics() cmp.Option {
+	return cmp.Transformer("Sort", func(in []metricExtract) []string {
+		out := make([]string, 0, len(in))
+		seen := map[string]int{}
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply metrics/record_e2e_test.go:111]: Yeah, particularly for a test. But the mix of initialization is odd.

File: metrics/record_e2e_test.go [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+		for _, m := range in {
+			// Keep only the newest report for a key
+			key := m.Key()
+			if seen[key] == 0 {
+				out = append(out, m.String())
+				seen[key] = len(out) // Store address+1 to avoid doubling first item.
+			} else {
+				out[seen[key]-1] = m.String()
+			}
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply metrics/record_e2e_test.go:120]: Done.

File: metrics/record_e2e_test.go [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+				collectorAddressKey:                 ocFake.address,
+				allowStackdriverCustomMetricsKey:    "true",
+				stackdriverCustomMetricSubDomainKey: servingDomain,
+				reportingPeriodKey:                  "1",
+			},
+		}
+	}
+
+	resources := []*resource.Resource{
+		{
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply metrics/record_e2e_test.go:150]: Done.

File: metrics/record_e2e_test.go [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+			Type: "revision",
+			Labels: map[string]string{
+				"project":  "p1",
+				"revision": "r1",
+			},
+		},
+		{
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply metrics/record_e2e_test.go:157]: Done, throughout.

**Comment by evankanderson** [comment]
Looking into the new segv...

**Comment by evankanderson** [comment]
```
$ (base) evana-a01:pkg evana$ go test ./metrics -run '^TestS[d]'
ok  	knative.dev/pkg/metrics	0.055s
$ go test ./metrics/... -run '^TestS[e]'
ok  	knative.dev/pkg/metrics	0.049s
$ go test ./metrics -run '^TestS[d-e]'
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x236ba92]

goroutine 6 [running]:
knative.dev/pkg/metrics.(*testExporter).ExportView(0xc00000efa0, 0xc0000be0f0)
	<autogenerated>:1 +0x32
go.opencensus.io/stats/view.(*worker).reportView(0xc00017e000, 0xc00000f400)
	/Users/evana/go/src/knative.dev/pkg/vendor/go.opencensus.io/stats/view/worker.go:357 +0x252
go.opencensus.io/stats/view.(*unregisterFromViewReq).handleCommand(0xc0001d87a0, 0xc00017e000)
	/Users/evana/go/src/knative.dev/pkg/vendor/go.opencensus.io/stats/view/worker_commands.go:98 +0xe7
go.opencensus.io/stats/view.(*worker).start(0xc00017e000)
	/Users/evana/go/src/knative.dev/pkg/vendor/go.opencensus.io/stats/view/worker.go:278 +0x1bb
created by go.opencensus.io/stats/view.init.0
	/Users/evana/go/src/knative.dev/pkg/vendor/go.opencensus.io/stats/view/worker.go:34 +0x68
FAIL	knative.dev/pkg/metrics	0.049s
FAIL
```

So there's something that changes test execution order here that breaks existing tests.

**Comment by evankanderson** [comment]
```
go test ./metrics -run '^(TestSdRecord|TestSetFactory)'
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x236ba92]

goroutine 19 [running]:
knative.dev/pkg/metrics.(*testExporter).ExportView(0xc000801100, 0xc0000b6690)
	<autogenerated>:1 +0x32
go.opencensus.io/stats/view.(*worker).reportView(0xc00017a000, 0xc000801460)
	/Users/evana/go/src/knative.dev/pkg/vendor/go.opencensus.io/stats/view/worker.go:357 +0x252
go.opencensus.io/stats/view.(*unregisterFromViewReq).handleCommand(0xc0008019e0, 0xc00017a000)
	/Users/evana/go/src/knative.dev/pkg/vendor/go.opencensus.io/stats/view/worker_commands.go:98 +0xe7
go.opencensus.io/stats/view.(*worker).start(0xc00017a000)
	/Users/evana/go/src/knative.dev/pkg/vendor/go.opencensus.io/stats/view/worker.go:278 +0x1bb
created by go.opencensus.io/stats/view.init.0
	/Users/evana/go/src/knative.dev/pkg/vendor/go.opencensus.io/stats/view/worker.go:34 +0x68
FAIL	knative.dev/pkg/metrics	0.048s
FAIL
```

**Comment by vagababov** [comment]
Something about how we setup globals?

**Comment by evankanderson** [comment]
Yeah, I think I've figured it out finally and may have reduced flakiness in TestMetricsExport, too.

File: metrics/exporter.go [hunk 9]
```diff
@@ -187,6 +187,10 @@ func isNewExporterRequired(newConfig *metricsConfig) bool {
 		return newConfig.collectorAddress != cc.collectorAddress || newConfig.requireSecure != cc.requireSecure
 	}
 
+	if newConfig.backendDestination == prometheus {
+		return newConfig.prometheusPort != cc.prometheusPort
+	}
+
 	return newConfig.backendDestination == stackdriver && newConfig.stackdriverClientConfig != cc.stackdriverClientConfig
 }
```
Comment (vagababov) [inline metrics/exporter.go:195]: return dest=promo && nc.pp != cc.pp || des == stack && nc.sCC != cc.sCC?

Comment (evankanderson) [reply metrics/exporter.go:195]: I might make this a switch, I worry about having a conditional with 4 boolean expressions being hard to maintain.

**Comment by evankanderson** [comment]
/close

Fixed this with #1957

**Pull request closed without merging**